### PR TITLE
Enable batch/v1beta1.CronJobs by default

### PR DESCRIPTION
**Pull request by soltysh**
This PR moves to CronJobs beta entirely, enabling `batch/v1beta1` by default.

Related issue: #41039 

@erictune @janetkuo ptal

```release-note
Promote CronJobs to batch/v1beta1.
```

### CODE REVIEW

**Comment by soltysh** [comment]
@deads2k for approval

File: cmd/kube-controller-manager/app/batch.go
Comment (janetkuo) [inline cmd/kube-controller-manager/app/batch.go]: Remove this line

File: cmd/kube-controller-manager/app/batch.go
Comment (janetkuo) [inline cmd/kube-controller-manager/app/batch.go]: `batch/v1beta1` cronjob clientsets are used in cronjob controller code, so `batch/v1beta1` has to be enabled?

File: cmd/kube-apiserver/app/server.go
Comment (janetkuo) [inline cmd/kube-apiserver/app/server.go]: What does this line do?

File: cmd/kube-controller-manager/app/batch.go
Comment (deads2k) [inline cmd/kube-controller-manager/app/batch.go]: Given client-go, it's basically impossible for a controller to tolerate two different API levels.  Are you sure this is doing what you expect?

**Comment by deads2k** [comment]
@lavalamp you should probably take a look at what's happening with these now.  The effects of lacking a hub version are pretty ugly.

File: cmd/kube-controller-manager/app/batch.go
Comment (deads2k) [reply cmd/kube-controller-manager/app/batch.go]: > Given client-go, it's basically impossible for a controller to tolerate two different API levels. Are you sure this is doing what you expect?

Yeah, if your controller is only going to work on the beta version (what it looks like), then you should only list the beta version as a gate.  Things don't go sideways until you switch to GA.

File: cmd/kube-controller-manager/app/batch.go
Comment (soltysh) [reply cmd/kube-controller-manager/app/batch.go]: Yeah, I initially thought that maybe be possible to work with both, but it's not. I forgot to update it. Will do.

File: cmd/kube-apiserver/app/server.go
Comment (soltysh) [reply cmd/kube-apiserver/app/server.go]: This makes sure CronJob is stored with batch/v1beta1. If this is not present current mechanics will try to find codec for CronJob in batch/v1, which will fail. This needs to be fixed, I've spoked with @deads2k about it earlier today, but that won't happen in 1.8 timeframe, I guess.

File: cmd/kube-controller-manager/app/batch.go
Comment (soltysh) [reply cmd/kube-controller-manager/app/batch.go]: Yup, I initially thought about working with both APIs, but failed. I forgot to update it.

**Comment by soltysh** [comment]
I've addressed comments.

File: pkg/controller/cronjob/cronjob_controller.go
Comment (janetkuo) [inline pkg/controller/cronjob/cronjob_controller.go]: Will this break HA clusters during upgrades? Not all masters are updated simultaneously: https://github.com/kubernetes/kubernetes/issues/51049#issuecomment-324784202

File: cmd/kube-apiserver/app/server.go
Comment (janetkuo) [reply cmd/kube-apiserver/app/server.go]: Does this change default storage version from batch/v2alpha1 to batch/v1beta1?

File: pkg/controller/cronjob/cronjob_controller.go
Comment (soltysh) [reply pkg/controller/cronjob/cronjob_controller.go]: I think this is slightly different usecase than what you're pointing to. There, the issue is about upgrading from beta1 to beta2, where both are enabled by default. Here, we are migrating from alpha to beta, iow. from non-default to a default enabled API. 
If I don't change controller to use the beta (enabled by default) cronjobs will not be working. Because you'll have the beta API available, but controller will rely on non-default API to be turned on. 
This will break HA clusters, but only those that explicitly enable the alpha version. I'll defer to @bgrant0607 or @smarterclayton for the decision but imho this should get in.

**Review by janetkuo — APPROVED** [review]

File: pkg/controller/cronjob/cronjob_controller.go
Comment (smarterclayton) [reply pkg/controller/cronjob/cronjob_controller.go]: We don't support HA alpha upgrading to beta without interruption.  So someone on a cluster like that has already broken the glass.

**Comment by smarterclayton** [comment]
Is there an issue in the features repo tracking this?

**Comment by janetkuo** [comment]
> Is there an issue in the features repo tracking this?

https://github.com/kubernetes/features/issues/19?

**Comment by janetkuo** [comment]
> *No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

@soltysh would you add the issue number to the PR body so that it can get approved?

**Comment by soltysh** [comment]
> @soltysh would you add the issue number to the PR body so that it can get approved?

Done.

**Comment by shyamjvs** [comment]
This PR is breaking presubmit e2e's due to the group version not being updated at a few places.
https://github.com/kubernetes/kubernetes/issues/51692

**Comment by justinsb** [comment]
If this PR breaks/impacts HA clusters, can we expand the release note to be a little more instructive please? cc @jdumars

**Comment by soltysh** [comment]
> If this PR breaks/impacts HA clusters, can we expand the release note to be a little more instructive please? cc @jdumars

I'll sync with @kubernetes/sig-docs-maintainers about it.

**Comment by jdumars** [comment]
Which SIG should this be assigned to?

**Comment by soltysh** [comment]
@jdumars I'm currently updating the release notes, I'll ping you for a double check. It's sig-apps

**Pull request merged**
